include/action.h

```diff
@@ -2,19 +2,48 @@
 #ifndef __LABWC_ACTION_H
 #define __LABWC_ACTION_H
 
-struct server;
+/* Forward declare arguments */
 struct view;
+struct server;
+struct wl_list;
+struct wlr_box;
 
 struct action {
-	uint32_t type;
-	char *arg;
-	struct wl_list link;
+	struct wl_list link;  /* struct keybinding.actions,
+	                       * struct mousebinding.actions,
+	                       * struct menuitem.actions */
+
+	uint32_t type;        /* enum action_type */
+	struct wl_list args;  /* struct action_arg.link */
 };
 
-struct action *action_create(const char *action_name);
-void action_list_free(struct wl_list *action_list);
+struct action_arg {
+	struct wl_list link;  /* struct action.args */
 
+	/**
+	 * FIXME: Do we want to replace `name` by a plain uint32_t and have
```

Comment:
Don't mind. action_run() isn't a function that gets called that often, so the tiny overhead of a strcmp()s probably pales into insignificance :smile: